### PR TITLE
asset-rate pallet: box asset kind parameter

### PR DESCRIPTION
**Pull request by muharem**
The `AssetKind` type parameter of a dispatchable, defined by the user, might be large — like `xcm::MultiLocation`. To prevent inflating the size of the `Call` type, we `Box` it.

This changes required for https://github.com/paritytech/polkadot-sdk/pull/1333

### CODE REVIEW

File: substrate/frame/asset-rate/src/lib.rs [hunk 1]
```diff
@@ -155,18 +156,18 @@ pub mod pallet {
...
 			origin: OriginFor<T>,
-			asset_kind: T::AssetKind,
+			asset_kind: Box<T::AssetKind>,
 			rate: FixedU128,
 		) -> DispatchResult {
 			T::CreateOrigin::ensure_origin(origin)?;
 
 			ensure!(
-				!ConversionRateToNative::<T>::contains_key(asset_kind.clone()),
+				!ConversionRateToNative::<T>::contains_key(*asset_kind.clone()),
```
Comment (ggwpez) [inline substrate/frame/asset-rate/src/lib.rs:165]:
```suggestion
				!ConversionRateToNative::<T>::contains_key(asset_kind.as_ref()),
```
No cloning needed.

File: substrate/frame/asset-rate/src/lib.rs [hunk 2]
```diff
@@ -155,18 +156,18 @@ pub mod pallet {
...
 				Error::<T>::AlreadyExists
 			);
-			ConversionRateToNative::<T>::set(asset_kind.clone(), Some(rate));
+			ConversionRateToNative::<T>::set(*asset_kind.clone(), Some(rate));
```
Comment (ggwpez) [inline substrate/frame/asset-rate/src/lib.rs:168]: ditto

**Review by ggwpez — APPROVED** [review]
There are some un-needed clones, but looking good in general.

**Comment by muharem** [comment]
@ggwpez can you look again please.
I also updated traits `ConversionFromAssetBalance` and `ConversionToAssetBalance` accept `AssetId` as ref to avoid extra clone.

**Comment by muharem** [comment]
probably better to have separate PR for the traits

**Comment by ggwpez** [comment]
> @ggwpez can you look again please. I also updated traits `ConversionFromAssetBalance` and `ConversionToAssetBalance` accept `AssetId` as ref to avoid extra clone. I will update the PR title if looks good to you.

Historically the converters use instances and not references, eg like in the [`Convert` trait](https://github.com/paritytech/polkadot-sdk/blob/22393d51e4a90799fd2541b696e413eab0a9e692/substrate/primitives/runtime/src/traits.rs#L563).  
I think it makes sense to keep that, since the functions are called `to_asset_balance`, which indicated consumption of the argument (as compared to `as_asset_balance`).

**Comment by muharem** [comment]
@ggwpez reverted

**Review by juangirini — APPROVED** [review]

**Comment by muharem** [comment]
@ggwpez can you reapprove if looks good, I did reset the previous approve from you.

**Review by ggwpez — APPROVED** [review]

**Comment by Polkadot-Forum** [comment]
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/polkadot-release-analysis-v1-2-0/4451/1

**Pull request merged**